### PR TITLE
add bizStateChangeLog

### PR DESCRIPTION
**Pull request by gaosaroma**
### Motivation

排查Biz状态变更问题更方便

### Modification

添加 bizStateChangeLog

### result
Ark Biz: biz1:0.0.1-SNAPSHOT, classloader: sun.misc.Launcher$AppClassLoader@18b4aac2, current state: unresolved, history states: [date: 2024-03-28 11:13:35.310, to resolved, date: 2024-03-28 11:13:35.313, to activated, date: 2024-03-28 11:13:35.314, to deactivated, date: 2024-03-28 11:13:35.318, to unresolved]

image
image

## Summary by CodeRabbit

- **New Features**
	- Enhanced business component state management with the ability to log and retrieve state change history.

- **Refactor**
	- Improved handling of state changes in business components, ensuring more detailed and reliable state management.

- **Tests**
	- Expanded test coverage to include new state logging and transition functionalities for business components.

### CODE REVIEW

**Comment by lvjing2** [comment]
@coderabbitai review

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/model/BizModel.java [hunk 1]
```diff
@@ -392,9 +402,16 @@ public Map<String, String> getAttributes() {
...
 
+    @Override
+    public CopyOnWriteArrayList<BizStateChangeInfo> getBizStateChangeLogs() {
+        return bizStateChangeLogs;
+    }
+
     @Override
     public String toString() {
-        return "Ark Biz: " + getIdentity();
+        String classloaderTag = classLoader == null ? "null" : classLoader.toString();
```
Comment (lvjing2) [inline sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/model/BizModel.java:412]: 不需要主动转换，直接打印即可

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/model/BizModel.java [hunk 2]
```diff
@@ -392,9 +402,16 @@ public Map<String, String> getAttributes() {
...
+        return "Ark Biz: " + getIdentity() + ", State: " + bizState + ", classloader: "
+               + classloaderTag + ", changeLogs: " + bizStateChangeLogs;
```
Comment (lvjing2) [inline sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/model/BizModel.java:414]: 按照如下顺序：
1. ArkBiz Identity
2. ClassLoader
3. Current State
4. History States

File: sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/model/BizInfo.java [hunk 3]
```diff
@@ -117,4 +120,27 @@ public interface BizInfo {
...
+
+    /**
+     * get getBizStateChangeLog
+     * @since 2.2.9
+     * @return java.util.concurrent.CopyOnWriteArrayList<com.alipay.sofa.ark.spi.model.BizInfo.BizStateChangeInfo>
+     */
+    CopyOnWriteArrayList<BizStateChangeInfo> getBizStateChangeLogs();
+
+    class BizStateChangeInfo {
+        private final Date     changeTime;
```
Comment (lvjing2) [inline sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/model/BizInfo.java]: Date 无法解决时区问题

**Comment by lvjing2** [comment]
最后截图展示下显示的效果

File: sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/model/BizInfo.java [hunk 4]
```diff
@@ -117,4 +120,27 @@ public interface BizInfo {
...
+            this.changeTime = changeTime;
+            this.state = state;
+        }
+
+        @Override
+        public String toString() {
+            String date = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS").format(changeTime);
+            return String.format("date: %s, to %s", date, state);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (lvjing2) [reply sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/model/BizInfo.java]: 棒

File: sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/model/BizInfo.java (hunk 3 above)
Comment (gaosaroma) [reply sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/model/BizInfo.java]: > Date 无法解决时区问题

sdf.setTimeZone(TimeZone.getDefault()); 可以么？

File: sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/model/BizInfo.java [hunk 5]
```diff
@@ -117,4 +123,34 @@ public interface BizInfo {
...
+        public BizStateChangeInfo(Date changeTime, BizState state) {
+            this.changeTime = changeTime;
+            this.state = state;
+        }
+
+        @Override
+        public String toString() {
+
+            String date = sdf.format(changeTime);
+            return String.format("date: %s, to %s", date, state);
```
Comment (lvjing2) [inline sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/model/BizInfo.java:153]: String.format("state change to %s at %s", state.toString(), date);

Comment (gaosaroma) [reply sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/model/BizInfo.java]: state 可以不显式调用 toString() 方法

**Review by lvjing2 — APPROVED** [review]
LGTM

**Pull request merged**